util/src/main/java/io/kubernetes/client/Discovery.java

```diff
@@ -60,16 +63,18 @@ public Set<APIResource> findAll() throws ApiException {
     return allResources;
   }
 
-  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion)
-      throws ApiException {
+  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion) {
     return findAll(group, versions, preferredVersion, "/apis/" + group + "/" + preferredVersion);
   }
 
-  public Set<APIResource> findAll(
-      String group, List<String> versions, String preferredVersion, String path)
-      throws ApiException {
-    V1APIResourceList resourceList = resourceDiscovery(path);
-    return groupResourcesByName(group, versions, preferredVersion, resourceList);
+  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion, String path) {
+    try {
+      V1APIResourceList resourceList = resourceDiscovery(path);
+      return groupResourcesByName(group, versions, preferredVersion, resourceList);
+    } catch (ApiException e) {
+      LOGGER.warn("Api {} returns a {} code, all api resources managed by this api cannot be discovered", path, e.getCode(), e);
```

Comment:
`initModelMap()` is always called statically:

Reply:
Also, concretely, the suggestion is not to ignore _all_ ApiExceptions, just to ignore this specific exception using the HTTP code/message to detect it.

Reply:
> `initModelMap()` is always called statically:
> 
> https://github.com/kubernetes-client/java/blob/c8923e6c4c2492ed9f44688813089d4663be4da9/util/src/main/java/io/kubernetes/client/util/ModelMapper.java#L74
> 
> So that still should work correctly.

Unfortunatly, the static initModelMap is not enougth to make extended client work without the discovery. This method does'nt init vars classesByGVR, and as demonstrated earlier, when this vars is not set, the extended client will lead to throwing another exception.

> Also, concretely, the suggestion is not to ignore _all_ ApiExceptions, just to ignore this specific exception using the HTTP code/message to detect it.

I ignore all exception because my goal is to simulate the kubectl command behavior. This command just print an error message to alert users that all api-resources cannot be listed when an api is in error (whatever the error is). All other apis are present.

Here an example of kubectl command returns: 
```
$ kubectl api-resources
NAME                              SHORTNAMES   APIVERSION                             NAMESPACED   KIND
bindings                                       v1                                     true         Binding
...
volumeattachments                              storage.k8s.io/v1                      false        VolumeAttachment
error: unable to retrieve the complete list of server APIs: bug.reproduce.java-client/v1alpha1: the server is currently unable to handle the request
```
I updated the warning message to be more like the one returned by the kubectl command.

I took the initiative of checking our respective implementations by running the program locally, and I confirm my previous analysis. Maybe if you're not convinced, you can check it locally too.

Reply:
Thanks for doing the analysis.

The problem here is that the `Discovery` class isn't just for Kubectl, it needs to be general purpose for any client, and if we catch an exception in the `Discovery` class then we are obfuscating an error for a person who wants to use the `Discovery` class outside of `Kubectl`

So emulating `kubectl` behavior inside of the `Kubectl` package is 100% fine, but other utility classes can't eat error messages.

Additionally, if you just catch the exception inside here: https://github.com/kubernetes-client/java/blob/master/util/src/main/java/io/kubernetes/client/util/ModelMapper.java#L269 and return an empty Set I believe that it will have the same effect as adding the catch where you added it.

Reply:
> I think that if you catch the exception here: https://github.com/kubernetes-client/java/blob/master/util/src/main/java/io/kubernetes/client/util/ModelMapper.java#L269 and return an empty Set I believe that it will have the same effect as adding the catch where you added it.

An empty Set at this location will not call addModelMap and will lead to an exception thrown by the Kubectl client as already explained. Do you agree with that?

> The problem here is that the `Discovery` class isn't just for Kubectl, it needs to be general purpose for any client, and if we catch an exception in the `Discovery` class then we are obfuscating an error for a person who wants to use the `Discovery` class outside of `Kubectl`
> 
> So emulating `kubectl` behavior inside of the `Kubectl` package is 100% fine, but other utility classes can't eat error messages.

Ok, I didn't understood that. So, in my opinion we have two solutions to fix this issue:
- Create a new checked exception, which contains a reference to all dicovered api resources, and will be thrown by discovery class. To ensure that the api contract is not broken by adding this new exception, it will extend ApiException. In library's code, this exception will be catched explicitly with api resources got from it. For user point of view, they will get an ApiException like before change.
- Update ModelMapper's methods to make them failover to prebuild metadata. All informations needed for this are not available in prebuild metadata https://github.com/kubernetes-client/java/blob/7f957585fe8fba9278c2a103a9f97cf6d1acbb9b/util/src/main/java/io/kubernetes/client/util/ModelMapper.java#L329-L331 We need to update KubernetesType to add a method isNamespaced, and update all implementation as a result. Personally, this solution is not my favorite one, I don't checked the impact of this modification, and it's pretty huge change.

Reply:
Thanks for the additional thought here. I'm ok with adding an adding a specialized checked exception here to handle this case.

I think it is also valuable to implement the fallback in the ModelMapper that you describe (I believe that this is what the Golang client does in `kubectl`) but I agree that it is extra work for this specific problem.

Reply:
I commited changes, would you please check it?